bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogger.java

```diff
@@ -1078,25 +1130,26 @@ private Header getHeaderForLogId(long entryLogId) throws IOException {
         }
     }
 
+
     private BufferedReadChannel getChannelForLogId(long entryLogId) throws IOException {
-        BufferedReadChannel fc = getFromChannels(entryLogId);
-        if (fc != null) {
-            return fc;
+        BufferedReadChannel brc = getFromChannels(entryLogId);
+        if (brc != null) {
+            // increment the refCnt
+            logid2FileChannel.get(entryLogId).retain();
+            return brc;
         }
+
         File file = findFile(entryLogId);
         // get channel is used to open an existing entry log file
         // it would be better to open using read mode
-        FileChannel newFc = new RandomAccessFile(file, "r").getChannel();
-        FileChannel oldFc = logid2FileChannel.putIfAbsent(entryLogId, newFc);
-        if (null != oldFc) {
-            newFc.close();
-            newFc = oldFc;
-        }
+        FileChannel fc = new RandomAccessFile(file, "r").getChannel();
+        logid2FileChannel.put(entryLogId, new ReferenceCountedFileChannel(fc));
```

Comment:
the logic should be similar as before. you need to use `putIfAbsent` for concurrent operations.